fix: Add username validator for LDAP storage provider to handle speci…

RomeoIndiaJulietUniform opened the pull request:
Hii @jonkoops @hokuda thanks a lot for assigning me this. As discussed I have made the changes by importing the **UsernameProhibitedCharactersValidator** and implementing it in the **importUserFromLDAP** method.  

Please tell me if I need to do anything more.  Thanks :)

Signed-off-by: Riju Mondal <mailrijumondal@gmail.com>

Closes: #32640

RomeoIndiaJulietUniform committed:
fix: Add username validator for LDAP storage provider to handle special characters

Signed-off-by: Riju Mondal <mailrijumondal@gmail.com>

jonkoops reviewed (approved):
Looks great, thank you for implementing this!

rmartinc commented:
@hokuda @RomeoIndiaJulietUniform @jonkoops Looking the PR I'm not totally sure that we should avoid any import of users with invalid characters from ldap. The reason those users are invalid in keycloak internal DB (no ldap provider) is the User Profile. The User Profile by default contains the username attribute with validations in lenght, username-prohibited-characters and up-username-not-idn-homograph. But if you delete the last two validators you can create a username with invalid characters in keycloak with no issues. So it's a configuration thing and not something compulsory at code level like this PR is doing for LDAP. This fix would be more about what to do if the user imported from ldap (or any other user provider) does not fulfill the UP validation (at any level, not just username). AFAIK right now user providers avoid the UP and the users in the external storage are valid for keycloak.

This is for the @keycloak/core-iam team so better if we wait for them to validate the PR.

mposolda commented:
+1 to @rmartinc . I also don't think that the `UsernameProhibitedCharactersValidator` should be hardcoded this way. The point of User-profile is exactly the ability to have various attribute restrictions configurable rather than hardcoded.

Maybe something to doublecheck, but maybe we check if we can have something like `Username mapper` for the LDAP, which can add some flexibility how to map the username from LDAP and eventually add some validations to that? We have the username mapper also for `IdentityProviderMapper` SPI, so maybe something similar can be for LDAP too? This will allow to remove option `Username LDAP attribute` directly from the LDAP provider and use the mapper setting instead, which will allow to remove some options from the provider itself. However it will require some migration if we go this way...

Another point: There is interface `UserProfileDecorator`, which allows to "decorate" the user profile configuration for the specific user storage provider (interface is implemented for LDAP). So maybe there is a way to use that one (maybe together with the username mapper)?

pedroigor commented:
We should be able to run validations on specific attributes at the provider level by taking into account whatever configuration you have for a given attribute.

When importing users we can build an UP object and run the validations for there. I also think this should be guarded by a flag (not sure if in the mapper or provider level) because this can introduce break existing deployments?

pedroigor commented:
> This fix would be more about what to do if the user imported from ldap (or any other user provider) does not fulfill the UP validation (at any level, not just username). AFAIK right now user providers avoid the UP and the users in the external storage are valid for keycloak.

For attributes the user can update (e.g.: first and last names) we can enforce they match the expectations from validators in the UP configuration by forcing the user to review their profile during authentication. However, it would be tricky to do the same for the username because this field would need to be editable and I'm not sure if it is always the case.

RomeoIndiaJulietUniform commented:
Hi @jonkoops, @rmartinc, @pedroigor, @mposolda,

Thank you all for the deep dive into the issue. I was thinking of implementing a factory pattern followed by a implementation of the various restriction that can be coupled to a UP object. 

Regarding the failed validation, if we don't fail the operation but instead flag it and send it back to the administrator, would that work?

I am really keen to work on this PR. It's thrilling to engage with something new and challenging. However, I need some time to fully understand the proper architecture. 

Do you have any specific implementation ideas in mind for this issue aligning with project structure?

RomeoIndiaJulietUniform committed:
fix: remove hardcoded username validation and use UP-based validation instead

Signed-off-by: Riju Mondal <mailrijumondal@gmail.com>

RomeoIndiaJulietUniform commented:
Hii @hokuda @jonkoops @pedroigor @rmartinc @mposolda I’ve removed the hardcoded validation that only checked prohibited characters and replaced it with validation based on all rules defined in the `User Profile (UP) configuration`. This uses the `UserProfileProvider,` which invokes the main `validate` method in `DefaultAttributes.java`.  

However, I still need clarification on the expected behavior when validation fails. The current implementation appends the suffix  `_warn_` to the username and proceeds with the import. Should the username instead be transformed into a valid one using a mapper, or is a different approach expected in this case?

RomeoIndiaJulietUniform committed:
fix: ldap import returns null in case of invalid username

Signed-off-by: Riju Mondal <mailrijumondal@gmail.com>

rmartinc reviewed (changes requested):
Hi @RomeoIndiaJulietUniform! I think that with this change there will be issues in the tests. Have you tried to run the ldap tests?

```
mvn clean install -DskipTests
cd testsuite/integration-arquillian/tests/base
mvn test -Pauth-server-quarkus -Dtest=org.keycloak.testsuite.federation.ldap.**.*Test
```

My feeling is that adding the UP check is not so easy. And probably we need to execute over all the attributes (not only the username).

RomeoIndiaJulietUniform commented:
Thanks, @rmartinc. I ran the LDAP tests and I’m seeing widespread failures.
I’m putting this change on hold for now until there is clarity on the expected behavior around applying User Profile validation during LDAP import (especially what should happen when an imported attribute fails validation).
